source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,337 @@ Codecs
+
+   Java type: java.util.List
+
+You can modify these mappings in your instance by specifying them in the
+``BsonTypeClassMap`` constructor. The following code snippet shows how
+you can replace the mapping for ``ARRAY`` in your ``BsonTypeClassMap``
+instance with the ``Set`` class:
+
+.. code-block:: java
+   :copyable: true
+
+   Map<BsonType, Class<?>> replacements = new HashMap<BsonType, Class<?>>();
+   replacements.put(BsonType.ARRAY, Set.class);
+   BsonTypeClassMap bsonTypeClassMap = new BsonTypeClassMap(replacements);
+
+   Class<?> clazz = bsonTypeClassMap.get(BsonType.ARRAY);
+   System.out.println("Java type: " + clazz.getName());
+
+This code outputs the following:
+
+.. code-block:: none
+   :copyable: false
+
+   Java type: java.util.Set
+
+For a complete list of the default mappings, refer to the
+:java-docs:`BsonTypeClassMap API documentation <apidocs/bson/org/bson/codecs/BsonTypeClassMap.html>`.
+
+For an example of how the ``Document`` class uses ``BsonTypeClassMap``, see
+the driver source code for the following classes:
+
+- :github:`DocumentCodecProvider <mongodb/mongo-java-driver/blob/master/bson/src/main/org/bson/codecs/DocumentCodecProvider.java>`
+- :github:`DocumentCodec <mongodb/mongo-java-driver/blob/master/bson/src/main/org/bson/codecs/DocumentCodec.java>`
+
+.. _codecs-custom-example:
 
 Custom Codec Example
 --------------------
 
-- Methods to override
-- Encode: Describe parameters: BsonWriter, EncoderContext
-- Decode: Describe parameters: BsonReader, DecoderContext
-- Example class, demonstrating implementing encode/decode. Perhaps something like this https://stackoverflow.com/questions/45083885/decode-document-to-a-java-class-using-custom-mongo-codec
-- Sample query to demonstrate decode results (of a field in a POJO) using custom codec
-- Sample insert and resulting fields of the document to demonstrate encode 
-- Example showing how to read/encode a document with unknown fields
+In this section, we show how you can implement ``Codec`` and ``CodecProvider``
+to define the encoding and decoding logic for a custom Java class. We also show
+how you can specify and use your custom implementations to perform insert
+and retrieve operations.
+
+The following code snippet shows our example custom class called ``Monolight``
+and its fields that we want to store and retrieve from a MongoDB collection:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/Monolight.java
+   :start-after: start class
+   :end-before: end class
+   :language: java
+   :dedent:
+
+This class contains the following fields, each of which we need to assign a
+``Codec``:
+
+- ``colorTemperature`` describes the color of the light and contains an
+  ``Integer`` value for which we use the ``IntegerCodec`` included in the
+  BSON library.
+- ``powerStatus`` describes whether the light is "on" or "off" for which we use
+  the :ref:`PowerStatusCodec <codecs-powerstatus-codec>` that converts specific
+  string values to BSON booleans.
```

Comment:
should these be ordered here the same as they are declared in the code, or does it matter?

Reply:
Good catch, will update.